proof_of_leadership/leadership_poseidon.circom

```diff
@@ -0,0 +1,301 @@
+    component bitifier = BLSNum2Bits_strict();
+
+    bitifier.in <== ticket;
+    bitifier.out[254] === 0;
+    for(var i=0; i<254; i++){
+        isLess.in[i] <== bitifier.out[i];
+    }
+    isLess.out === 0;
+
+
+    
+            // Compute the threshold
+    signal intermediate_value;
+    signal threshold;
+    intermediate_value <== t0 + t1 * value;
+    threshold <== intermediate_value * value;
+
+            // Ensure that the ticket is indeed 253 bits and that the ticket is winning
+    component isLess2 = BLSLessThan(253);
+    isLess2.in[0] <== ticket;
+    isLess2.in[1] <== threshold;
+    isLess2.out === 1;
+}
+
+
+template nullifier_computer(){
+    signal input note_nonce;
+    signal input nullifier_secret_key;
+    signal input value;
+    signal output nullifier;
+
+    component hash = hash_4_to_1();
+
+            //The b"coin-nullifier" Tag converted in F_p element (from bits with big endian order)
+    hash.in[0] <== 2016785505923014207119328528655730;
+    hash.in[1] <== note_nonce;
+    hash.in[2] <== nullifier_secret_key;
+    hash.in[3] <== value;
+
+    nullifier <== hash.out;
+}
+
+template commitment_computer(){
+    signal input note_nonce;
+    signal input nullifier_public_key;
+    signal input value;
+    signal output commitment;
+
+    component hash = hash_4_to_1();
+
+            //The b"coin-commitment" Tag converted in F_p element (from bits with big endian order)
+    hash.in[0] <== 516297089516239580383111224192495220;
+    hash.in[1] <== note_nonce;
+    hash.in[2] <== nullifier_public_key;
+    hash.in[3] <== value;
+
+    commitment <== hash.out;
+}
+
+template nonce_updater(){
+    signal input note_nonce;
+    signal input nullifier_secret_key;
+    signal output updated_nonce;
+
+    component hash = hash_4_to_1();
+
+            //The b"coin-evolve" Tag converted in F_p element (from bits with big endian order)
+    hash.in[0] <== 120209783668687835891529317;
+    hash.in[1] <== note_nonce;
+    hash.in[2] <== nullifier_secret_key;
+    hash.in[3] <== 0;
+
+    updated_nonce <== hash.out;
+}
+
+template membership_checker(){
+    signal input leaf;                      //The note commitment
+    signal input root;                      //The root of the Merkle Tree (of depth 32)
+    signal input index[32];                 //Position of the note commitment in bits in big endian
+    signal input node[32];                  //Complementary hashes
+
+    component hash[32];
+
+    for(var i=0; i<32; i++){
+        hash[i] = hash_2_to_1();
+    }
+
+    hash[0].in[0] <== leaf - index[31] * (leaf - node[0]);
+    hash[0].in[1] <== node[0] - index[31] * (node[0] - leaf);
+
+    for(var i=1; i<32; i++){
+        hash[i].in[0] <== hash[i-1].out - index[31] * (hash[i-1].out - node[i]);
+        hash[i].in[1] <== node[i] - index[31] * (node[i] - hash[i-1].out);
+    }
+
+    root === hash[31].out;
+
+}
```

Comment:
So, I thought merkle proofs required L/R tags for each node in the merkle proof.

Here we assume that the left hand side of each node is the "spine" of the path. Are you assuming that the merkle proof nodes are flipped to maintain that property? This seems safe to my eyes, is this a standard optimization to use?

Reply:
No the L/R tag is in the index signal.
But since conditional statements like if are not possible (since the circuit must always do the same whatever the inputs),
just use the trick L <- A - index (A-B) and R <- B - index(B-A) then depending on index we can say if A is on the left or on the right.

If A is a left node we input index = 0 then L <- A and R <- B.
If A is a right node we input index = 1 then L <- A - (A- B) = B and R<- B - (B - A) = A

Reply:
But we could also adopt your optimization as well (since the path is also given through the inputs by the prover)